### ISSUE TITLE
Proposal: new ReactionDSL based on SciCompDSL

### ISSUE THREAD

**Issue by AleMorales** [SciML/DiffEqBiological.jl#27]
I thought it would be an interesting to start building in parallel a new DSL for biological reactions on top of [SciCompDSL.jl](https://github.com/JuliaDiffEq/SciCompDSL.jl). I have some ideas about it an a draft of the data structures/operators that could be used, but I would like to throw it around first.

First, the system I have in mind so far would have the following features:

* Each reaction must have a rate equation which is just an arithmetic expression like the ones supported in SciCompDSL. This rate equation must explicitly use the reactants in the reaction. This may be cumbersome for simple mass action models but it allows to be more general and even go beyond mass action (e.g. Goldman–Hodgkin–Katz flux equation for ion transport)
* There is no need for directionality in the arrows to specify the reaction as the rate equation already includes the logic. So one arrow suffices for all types of reactions.
* Reactants can be defined to be inside compartments as well as reactions. This allows to simulate transport across compartments.
* Reactions may be mixed with explicit differential equations, intermediate variables, etc (i.e. what is already available at SciCompDSL). The ReactionDSL will simple lower the reactions into intermediate expressions and differential equations and leave the rest untouched.
* A reactant may also appear in an explicit differential equations, intermediate variable, etc and the lowering of reactions will just be merged with what is already there.

Some types are needed:

* `Compartment`: This can be lower to a `Parameter` or  a `DependentVariable` based on whether it is constant or changes during simulation (e.g. due to cell growth). A compartment does not have to be a volume, it could be a surface too. The role of compartment is to perform transformations when a reactant and reaction are not defined on the same compartment (see above).
* `Species`: This means "chemical species", not the biological concept. This is lowered to a `DependentVariable`. Its differential equation is generated by adding up all the rate equations of the reactions it is involved, modified by stoichiometry and compartment (when different).
* `Reactant`: The basic component of a reaction. This is basically a `Species` with an stochiometry. In most cases the stoichiometry would be a constant integer, but non-integer stoichiometries exist (e.g. check ATP synthase) and it may also be an idea to make it a parameter for people designing biological systems that don't exist yet (e.g. synthetic biology).
* `Reaction` is just a list of reactants (organized as substrates and products) and a compartment.

Operators can be overloaded to generate the right types. For example:

```
(2p + 2q → 2r => C) == k*(2p*2q - 2r/Keq)
```

The `→` can be used to generate the `Reaction` by analyzing the AST generated on each side (i.e. `Species` would behave like `Variables` in SciCompDSL).
`=>` may be used to indicate the compartment of the reaction (by default the `Reaction` would  get the compartment of the first `Reactant` so this is not always needed). I would really like to use `in` instead but then we would need to use parenthesis due to operator precedence.
The `==` would generate an `Operation` as in SciCompDSL, with a `Reaction` as first argument. Because of operator precedence, the parenthesis are required. A function would then process all the `Operation`s that have reactions and lower them. The rest of `Operation` in the model will be ignored. In the end, the lowered and ignored `Operation`s are passed to SciCompDSL to further processing into an ODE, SDE, etc...

I would love if ReactionDSL would eventually have all the features that DiffEqBiological has already and to make the transition as painless as possible, hence my posting of this message here. In particular, I am not sure whether the "single arrow approach + explicit use of reactants in rate equation" will be seen as step backwards for most people (though, again, not all biological models follow mass action). Also, I have only built ODEs in the past, so I am not sure how this would address the other types of models.

@korsbo @Gaussia

**Comment by korsbo** [comment]
Me and @gaussia have already started talking about jacking DiffEqBiological into the SciCompDSL. We said that we may wait until SciCompDSL has stabilised a bit, but I think that this should be done eventually. 

As to the particulars, I can't give you much feedback at the moment. I would first have to get a better grasp of what SciCompDSL is developing into. I do intend to read up on this, but it may take a little time. 

However, my initial thoughts (subject to revision upon better understanding) are:
- Orthogonality is good. This package should depend on SciCompDSL, and we should be careful not to implement similar things in different packages. The same argument holds for ParameterizedFunctions.jl. The low level stuff (symbolic calcs, multi-cell functionality, etc) should exist in only one place, and the specific DSL packages should import rather than reimplement. Also, in the long term, I do not think it makes sense to maintain two packages which both supply some reaction DSL.

- I am slightly hesitant to enforce a more verbose DSL. I like the compactness of both `@reaction_network` and `@ode_def`. While I agree that these DSLs should be extended, I would hope that we can create clever defaults so that these extensions do not necessarily increase verbosity (much).  

- I am a massive fan of the efforts to get nice DSLs which are compatible with spatial models. 


I'll read up on the SciCompDSL and hopefully increase my understanding to a point where I can be more useful. Also, if you could create a sample of how you envision the DSL would look like, that may help me get a more clear picture of what you mean.

**Comment by AleMorales** [comment]
Thanks for the detail reply. 

An example of syntax for a simple model would be this:

```julia
using SciCompDSL
using ReactionDSL
@Species x, y, xy, z # Equivalent to x = Species(:x); y = Species(:y)...
@IVar t
reactions = [x + y → xy == 2,
                    xy → z == 1]
de = ReactionDES(reactions, t)
f = DiffEqFunction(de)
```

And `f` can already be fed to the solver (have not tried yet though). Note that half of the functionality in the example is provided by SciCompDSL. ReactionDSL would only add some new types (e.g. `Species`) and operators (`→`), and functions that lower the data structure to ODE, SDE, etc (`ReactionDES`). 

In this example the `→` creates a formula that keeps track of the reactants and their stoichiometry (each type of arrow would have a different Formula type). The `==` creates an `Operation` with the formula and its rate equation as argument (the latter is handled by SciCompDSL). `ReactionDES` would then construct the derivatives for the different chemical species using the formulae and their rate equations. After that, it is all SciCompDSL.

An advantage of this approach is that I can include statements describing algebraic or differential equations and `ReactionDES` will just ignore them and pass them untouched. So it would be really easy to make hybrid models that use different formalisms for different parts of the model (e.g. reactions for metabolism, differential equations for biophysics).

**Comment by ChrisRackauckas** [comment]
>An advantage of this approach is that I can include statements describing algebraic or differential equations and ReactionDES will just ignore them and pass them untouched. So it would be really easy to make hybrid models that use different formalisms for different parts of the model (e.g. reactions for metabolism, differential equations for biophysics).

Getting that right automatically takes a lot of care which is why it was avoided in the current DSL, but the SciCompDSL IR has enough information to be able to choose the right representation to do this. 

>In this example the → creates a formula that keeps track of the reactants and their stoichiometry (each type of arrow would have a different Formula type). The == creates an Operation with the formula and its rate equation as argument (the latter is handled by SciCompDSL). ReactionDES would then construct the derivatives for the different chemical species using the formulae and their rate equations. After that, it is all SciCompDSL.

This doesn't cover the full span of what's allowed in the DSL though since it doesn't incorporate syntax for choosing regulation functions. This needs some bikeshedding. 

>I am slightly hesitant to enforce a more verbose DSL. I like the compactness of both @reaction_network and @ode_def. While I agree that these DSLs should be extended, I would hope that we can create clever defaults so that these extensions do not necessarily increase verbosity (much).

I think it's better to think of what's evolving as a scientific computing intermediate representation (IR) for a scientific computing compilation process. Like LLVM, the IR is being developed so it's human legible, but the main goal is to be a comprehensive format for scientific computing models in order to do model manipulations before compilation (it's compilation process is simply generating Julia functions). Macros can be used to write to the IR, so the `@reaction_network` macro could actually be the same interface but create a different internal format which is more amenable to transformations (the SciCompDSL IR). If we get jumps right in the SciCompDSL IR, then presumably this macro only needs to build the jump equation, and the conversion to ODEs/SDEs could occur as IR transformations instead of being something bio-specific.

**Comment by korsbo** [comment]
This looks very good. The hybrid bit is fantastic. 

Am I understanding it correctly when I think that your example would result in an equivalent system as the current:

```julia
@reaction_network test begin
    2, x + y --> xy
    1, xy --> z
end 
```
and that you would feed in parameters `r_1` and `r_2` like
```julia
reactions = [x + y → xy == r_1,
             xy → z == r_2]
```

> - There is no need for directionality in the arrows to specify the reaction as the rate equation already includes the logic. So one arrow suffices for all types of reactions.

Why would you not allow binding/unbinding by
```julia
reactions = [x + y ↔ xy == (r_bind, r_unbind)]
```
?

**Comment by korsbo** [comment]
> so the `@reaction_network` macro could actually be the same interface but create a different internal format

I understand that we could just keep the current front-end and plug it in to the new backend, but I also see why it would make sense to extend the front-end too. 

Currently, the order of the parameters in `@reaction_network` is well defined, but not the order of the variables (which are inferred, not specified). We also have no way of dealing with spatial stuff. This could of course be added to DiffEqBiological. But if we instead could just "inherit" it from SciCompDSL, that would be better. I therefore like the hybrid part, any functionality added to the SciCompDSL expressions would instantly be available here too.

**Comment by AleMorales** [comment]
> There is no need for directionality in the arrows to specify the reaction as the rate equation already includes the logic. So one arrow suffices for all types of reactions.
> Why would you not allow binding/unbinding by

Sorry, I think my examples are getting mixed up. In my first post I was talking about a single type of arrow, which would always subtract the rate from the substrate and add it to the product (according to stoichiometry) but not try to build the rate equation assuming mass action or other kinds of  mechanism, but simply use the rate equation. I think this is what the open arrows in DiffEqBiological? With this type of arrow you could say

```
x + y -> z == k*(x*y - z/Keq)
```
or

```
x + y -> xy == kbind*x*y - krelease*xy
```

And it can represent both directions automatically. Well, at least I know this works with deterministic ODE, in other approaches may be not, I don't know.

**Comment by AleMorales** [comment]
> An advantage of this approach is that I can include statements describing algebraic or differential equations and ReactionDES will just ignore them and pass them untouched. So it would be really easy to make hybrid models that use different formalisms for different parts of the model (e.g. reactions for metabolism, differential equations for biophysics).
> 
> Getting that right automatically takes a lot of care which is why it was avoided in the current DSL, but the SciCompDSL IR has enough information to be able to choose the right representation to do this.

I agree, it is the responsibility of the modeler to make everything consistent and there is a limit to what can be checked on the IR. But the nice thing here is that ReactionDSL does not need to know about statements it does not care about (provided they all conform to a know API).

**Comment by korsbo** [comment]
Ok, I now think that I understand your syntax. You may run into problems with stochastic simulations. It's important to separate what rates belong to the forward or the backward reaction so that you can assign it the right amount of noise (or create jump propensities, etc.). In `@reaction_network` this is done by comma separation, but I think that you are doing it by sign alone.  Even though I can't think of a chemically valid case, one may imagine a reaction where the forward rate is regulated by the difference of some terms. 

I also understand what you meant by not assuming mass action and I am conflicted about it. I'm rather happy with the assumption that is made by `@reaction_network` -- mass action unless you specify otherwise (with the open arrows). I rely heavily on mass action, and removing this default would clutter up my model definitions. Would it be hard to retain this behaviour with the new syntax?

**Comment by AleMorales** [comment]
Yes I agree with arrows, there is plenty of them and it is best to give different defaults. About different rates per direction I am not sure at the moment, but we could have an array of expressions after == (so similar to the commas).

Note that one limitation is getting all operators to work in the right order, so some parentheses may be needed to get correct parsing. Hopefully that does not clutter too much.

**Comment by Gaussia** [comment]
Throwing myself into the discussion as well.

- The SciCompDSL is moving quite fast right now, at the point I have slightly lost track of exactly where it is. At some point I definitely think we should jack  `@reaction_network` into that to give it some additional stuff for free. When/If depends a bit on when said SciCompDSL is more or less ready and we know what it will be (or on my part, me fully reading through what it is doing/will do).
- I am quite fond of the simply syntax of the current `@reaction_network`. While some functionality admittedly would require more input (like determining order of species) I'd like to keep such input optional. 
- On the same issue I think that the default arrows should use Mass Kinetics since I think this is the most common case. Then reactions not depending on Mass Kinetics use special arrows/notation.
- Making it possible to have parameter defined stoichiometry (like `N*X --> XN` with `N` a parameter) is on my to-do list. There is no real reason this should not be in.
- Hybrid systems were only part of the models are a biochemical reaction network is definitely something I would like to see. I have encountered situations myself were I wanted to build such a model.
- Spatial models have been the natural "next step" for the DSL (Admittedly this is no great speciality of mine). It probably would require some good thinking of what the best way for a modeller to input spatial parts would be. I'd be interested to see how we possibly could use the SciCompDSL to get this for free.

In summary I think we should try and free-ride as much as possible on the SciCompDSL, however I would also prefer to be able to keep (at least the option of) an as simple syntax as possible.

**Comment by AleMorales** [comment]
In order to have more concrete discussions and testing how far these ideas can go, I have implemented a first draft of how this DSL could look like at [ReactionDSL.jl](https://github.com/AleMorales/ReactionDSL.jl). See tests folder for examples of syntax and features.

For simplicity I have implemented only one type of arrow and single rate, but adding other types of arrows should be straightforward (see the issue AleMorales/ReactionDSL.jl#1), as well as having a different rate per direction (AleMorales/ReactionDSL.jl#3).

There are however some more annoying issues regarding syntax (because of operator precedence) and storing additional metadata inside SciCompDSL variables (e.g. for implementing compartments)... But any solution to these problem should simplify the syntax, not make it more complex.
